proxy_await.c

```diff
@@ -286,6 +286,16 @@ int mcplib_await_return(io_pending_proxy_t *p) {
                         valid = false;
                     }
                     break;
+                case AWAIT_FASTGOOD:
+                    if (p->client_resp->status == MCMC_OK) {
+                        // End early on a hit.
+                        if (p->client_resp->resp.code != MCMC_CODE_END) {
+                            aw->wait_for = 0;
```

Comment:
Curious if `cleanup` will always eventually end up being `true` for a FASTGOOD request even if we return early?
i.e., Is it guaranteed that every `io_pending_proxy_t` calls mcplib_await_return(), and so every request's state will be eventually cleaned up?

Reply:
cleanup is enforced by `aw->pending`. `aw->wait_for` is only for early return scenarios.

Reply:
Ack. I was just checking to see if `mcplib_await_return()` will always be called if `pending > 0` and `wait_for == 0`, but it sounds like it will.

Reply:
So far as I know you use it like that all the time so you'd notice if it were leaking.

`mcplib_await_return` is called for each pending IO that was dispatched unconditionally. internally it decides when to throw its associated client object back to work (ie `wait_for == 0`), and when all pending IO's are done the await object is reaped. The code is laid out a bit weird (it got refactored three times) but it's a state machine. I could rewrite it again to use a switch statement and be more clear but I don't like touching working code :)

It's already pending a refactor to reduce/remove the lua bits so I'll maybe get it then.

Reply:
To back up a bit, when calling `mcp.await()` in lua, it creates a high level await object and immediately dispatches all requests, each wrapped with their own IO objects. As those requests complete they are shipped back to the worker thread one at a time where the return code is executed. The return code checks the conditions of the responses for building the lua result table, when to dispatch the client back to work, and when to complete and reap the parent await object.

Reply:
> it creates a high level await object and immediately dispatches all requests, each wrapped with their own IO objects

Ack, thanks for that explanation- so to reiterate my understanding:-
- There are resources at 2 levels, the context for a request is at the `await` layer (or a `mcp_await_t`)
- The resources pertaining to a single IO is represented by `io_pending_proxy_t`

`io_pending_proxy_t` is guaranteed to call the callbacks once the IO is either complete, errors out or after some timeout. Once all the pending IOs call the callbacks, the await resources are cleaned up

Reply:
yup.